### PR TITLE
KT-45777: Set up artifact transform for incremental Kotlin compile

### PR DESCRIPTION
**Pull request by hungvietnguyen**

### CODE REVIEW

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/TasksProvider.kt [hunk 1]
```diff
@@ -87,14 +88,43 @@ internal open class KotlinTasksProvider {
...
+        }
+        configure(kotlinCompile, project, properties, compilation)
+        return kotlinCompile
+    }
+
+    companion object {
+        private const val TRANSFORMS_REGISTERED = "_kgp_internal_kotlin_compile_transforms_registered"
+    }
+
+    private fun registerTransformsOnceForKotlinCompile(project: Project) {
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/TasksProvider.kt:114]: Can you move configuration&transform creation to a companion object of `KotlinCompile.Configurator`?

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt [hunk 2]
```diff
@@ -426,6 +439,14 @@ abstract class KotlinCompile @Inject constructor(
...
     abstract val useClasspathSnapshot: Property<Boolean>
 
+    @get:Classpath
+    @get:Optional // Set if useClasspathSnapshot == true
+    abstract val classpathSnapshot: ConfigurableFileCollection
+
+    @get:OutputDirectory
+    @get:Optional // Set if useClasspathSnapshot == true
+    abstract val classpathSnapshotDir: DirectoryProperty
+
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt:449]: Would creating a `@Nested` input make this more elegant?

File: build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+sealed class ClasspathChanges : Serializable {
+
+    sealed class JVM : ClasspathChanges() {
+
+        sealed class ClasspathSnapshotEnabled : JVM() {
+
+            sealed class IncrementalRun : ClasspathSnapshotEnabled() {
+
+                class Available(val lookupSymbols: Collection<LookupSymbol>, val fqNames: Collection<FqName>) : IncrementalRun() {
```
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt]: Yup, I've cleaned up this class.

File: core/compiler.common/src/org/jetbrains/kotlin/name/FqName.java [hunk 4]
```diff
@@ -19,9 +19,12 @@
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.kotlin.utils.StringsKt;
 
+import java.io.Serializable;
 import java.util.List;
 
-public final class FqName {
+public final class FqName implements Serializable {
```
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply core/compiler.common/src/org/jetbrains/kotlin/name/FqName.java]: Yes - Done

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/TasksProvider.kt (hunk 1 above)
Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/TasksProvider.kt]: Done

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt (hunk 2 above)
Comment (hungvietnguyen) [reply libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt]: Done

**Comment by hungvietnguyen** [comment]
Rebased and resolved a few (trivial) merge conflicts.

File: build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+            out.writeInt(fqNames.size)
+            fqNames.forEach {
+                out.writeUTF(it.asString())
+            }
+        }
+
+        private fun readObject(ois: ObjectInputStream) {
+            val lookupSymbols = mutableListOf<LookupSymbol>()
+            repeat(ois.readInt()) {
```
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt]: Done

File: build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ */
+
+package org.jetbrains.kotlin.incremental
+
+import org.jetbrains.kotlin.name.FqName
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.io.Serializable
+
+sealed class ClasspathChanges : Serializable {
```
[the remark this answers is not in the document]
Comment (hungvietnguyen) [reply build-common/src/org/jetbrains/kotlin/incremental/ClasspathChanges.kt:17]: Ah yes - Done.

**Comment by hungvietnguyen** [comment]
Rebased and resolved a trivial merge conflict.

**Comment by nav-nav** [comment]
merged into master

**Comment by hungvietnguyen** [comment]
Awesome, thanks!

**Pull request closed without merging**